### ISSUE TITLE
some operations about the CAM_Module

### ISSUE THREAD

**Issue by junjieAI** [junfu1115/DANet#62]
in attention.py file, the code:
energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy
how do I transform this operation into tensorflow, thanks!

**Comment by junjieAI** [comment]
I've completed, thanks

**Comment by junjieAI** [comment]
@junfu1115 
by the way if I use tensorflow how do I update the parameters α and β,  thanks

**Comment by junfu1115** [comment]
@junjieAI Sorry, I am unfamiliar with tensorflow, the learned parameters α and β is initialized at the beginning

**Comment by junjieAI** [comment]
@junfu1115 yes, your paper mentioned that the parameter α and β are initialized as 0, and gradually learn to assign more weight, how about the process(whether is there a range from 0 to a defined value, and α, β are changes corresponding to the iterations. eg, when arrives to 10k iteration, α or β equal to 0.1 or some other value)? thanks

**Comment by lizleo** [comment]
@junjieAI Hi, how do you transform the operation and parameters into tensorflow, have you solved them? Could you please give me a hand? Please~

**Comment by junjieAI** [comment]
@lizleo 
the code as follow, but the trained model performance is not ideal, and I think my implementation also exists problems, if you get further progress, please remind me, thanks.
def danet(net_encoder, in_channels, planes, num_classes, gama, beta):
    # =================================== PAM ========================================
    # sa
    # conv5a
    feat1 = Head(net_encoder, in_channels, is_training=True, name='res5feature1')
    m_batchsize, height, width ,C= feat1.shape
    proj_query = PAM_Module(feat1, planes, name='pam1', transpose=True)
    proj_key = PAM_Module(feat1, planes, name='pam2')
    energy = tf.matmul(proj_query, proj_key)
    attention = tf.nn.softmax(energy)
    proj_value = PAM_Module_conv(feat1, planes, name='pam3', reshape=True)
    out = tf.matmul(proj_value, tf.transpose(attention,[0, 2, 1]))
    out = tf.transpose(tf.reshape(out, [m_batchsize, C, height, width]),[0, 2, 3, 1])
    sa_feat = gama*out + feat1
    # conv51
    sa_conv = PAM_Module_conv(sa_feat, planes, kernel_size=3, BN=True, Relu=True, name='pam_conv')
    # conv6
    sa_output = tf.nn.dropout(sa_conv, keep_prob=0.1)
    output = PAM_Module_conv(sa_output,512, num_classes,Out_Classes=True, name='pam_out')
    # =================================== PAM ========================================
    
    # =================================== CAM ========================================
    # conv5c
    feat2 = Head(net_encoder, in_channels, is_training=True, name='res5feature2')
    m_batchsize, height, width, C = feat2.shape
    proj_query = tf.reshape(feat2, [m_batchsize, C, -1])
    proj_key = tf.transpose(tf.reshape(feat2, [m_batchsize, C, -1]), [0, 2, 1])
    energy = tf.matmul(proj_query, proj_key)
    energy_init = tf.ones_like(energy)
    energy_new = tf.reduce_max(energy,-1, keepdims=True)
    energy_new = tf.multiply(energy_init, energy_new)
    energy_new = energy_new - energy
    attention = tf.nn.softmax(energy_new)
    proj_value = tf.reshape(feat2, [m_batchsize, C, -1])
    out = tf.matmul(attention, proj_value)
    out = tf.transpose(tf.reshape(out, [m_batchsize, C, height, width]), [0, 2, 3, 1])
    sc_feat = beta*out + feat2
    # conv52
    sc_conv = PAM_Module_conv(sc_feat, planes, kernel_size=3, BN=True, Relu=True, name='cam_conv')
    # conv7
    sc_output = tf.nn.dropout(sc_conv, keep_prob=0.1)
    sc_output = PAM_Module_conv(sc_output, 512, num_classes, Out_Classes=True, name='cam_out')

    feat_sum = sa_conv + sc_conv

    sasc_output = tf.nn.dropout(feat_sum, keep_prob=0.1)
    output = PAM_Module_conv(sasc_output, 512, num_classes, Out_Classes=True, name='pam_cam')
    return output

**Comment by lizleo** [comment]
@junjieAI Hello, I think your implementation of "energy_new" in the PAM is different from the original code. When you do "tf.multiply(energy_init, energy_new)", you will get max channel attention of each batch, but in the original code, the author did "torch.max(energy, -1, keepdim=True)[0].expand_as(energy)", and he chose the max channel attention of the first batch. I don't know why choose the first batch and which is right, but I think your codes are different. And I just do "energy_new = tf.reduce_max(energy, axis=[-1], keepdims=True)[0][tf.newaxis,:] - energy " to copy the original code.

**Comment by JiangYi0311** [comment]
In CAM_Module, what is meaning about the code "energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy"，this is not mentioned in the paper. thanks!

**Comment by Author123456** [comment]
@JiangYi0311, Hi Do you know why using that "energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy"? Seems that choose the least similarily across the channel? But why?

**Comment by way151** [comment]
@junfu1115 @Author123456 @JiangYi0311 Hi, I wonder whether you do some experiments on using `energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy` or just commenting it and using `softmax(energy)`. Would you pls showing me corresponding results if you have made such an ablation study? Thanks a lot!

**Comment by junfu1115** [comment]
> @junfu1115 @Author123456 @JiangYi0311 Hi, I wonder whether you do some experiments on using `energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy` or just commenting it and using `softmax(energy)`. Would you pls showing me corresponding results if you have made such an ablation study? Thanks a lot!

Ref https://github.com/junfu1115/DANet/issues/9

**Comment by way151** [comment]
> 
> 
> > @junfu1115 @Author123456 @JiangYi0311 Hi, I wonder whether you do some experiments on using `energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy` or just commenting it and using `softmax(energy)`. Would you pls showing me corresponding results if you have made such an ablation study? Thanks a lot!
> 
> Ref #9
> 
> Thanks, I understand it now. But I'm still confused by using `energy_new = max(energy) - energy` rather than `energy_new = energy - max(energy)`. I find the latter one is commonly used to prevent 'softmax overflow'. Does the former one mean that given two channels, e.g. i and j, if channel j is not similar to i, which means energy_{i,j} is small, then channel i should pay more attention to j, which means energy_new_{i,j} is big?

**Comment by mengce97** [comment]
> '. Does the former one mean that given two channels, e.g. i and j, if channel j is not similar to i, which means energy_{i,j} is small, then channel i should pay more attention to j, which means energy_new_{i,j} is big?

i think so, the more similarity pay less weights, this confused me

**Comment by Luffy03** [comment]
> In CAM_Module, what is meaning about the code "energy_new = torch.max(energy, -1, keepdim=True)[0].expand_as(energy)-energy"，this is not mentioned in the paper. thanks!

This confused me too. Have you found out why?

**Comment by chunbolang** [comment]
@mengce97 

Hi, I think your understanding is correct, but the weight of attention mechanism is learnable, that is, a negative(-) scaling factor would reverse this situation.
